tuta-sdk/rust/src/login/login_facade.rs

```diff
@@ -0,0 +1,273 @@
+use crate::crypto::key::GenericAesKey;
+use crate::entities::sys::{Session, User};
+#[mockall_double::double]
+use crate::typed_entity_client::TypedEntityClient;
+use crate::generated_id::{GENERATED_ID_BYTES_LENGTH, GeneratedId};
+use crate::login::credentials::Credentials;
+use crate::user_facade::UserFacade;
+use crate::util::{array_cast_slice, BASE64_EXT};
+
+/// Error that may occur during login and session creation
+#[derive(Error, Debug, uniffi::Error)]
+pub enum LoginError {
+    #[error("SessionIdError: {error_message}")]
+    SessionIdError {
+        error_message: String,
+    },
+    #[error("PassphraseError: {error_message}")]
+    PassphraseError {
+        error_message: String,
+    },
+    #[error("KeyError: {error_message}")]
+    KeyError {
+        error_message: String,
+    },
+    #[error("ApiCallError: {source}")]
+    ApiCallError {
+        #[from]
+        source: ApiCallError,
+    },
+}
+
+pub enum KdfType {
+    Bcrypt,
+    Argon2id,
+}
+
+/// Simple LoginFacade that take in an existing credential
+/// and returns a UserFacade after resuming the session.
+struct LoginFacade {
+    entity_client: Arc<TypedEntityClient>,
+}
+
+impl LoginFacade {
+    pub fn new(
+        entity_client: Arc<TypedEntityClient>,
+    ) -> Self {
+        LoginFacade {
+            entity_client,
+        }
+    }
+
+    async fn derive_user_passphrase_key(&self, kdf_type: KdfType, passphrase: &str, salt: [u8; 16]) -> Aes256Key {
+        match kdf_type {
+            KdfType::Argon2id => {
+                generate_key_from_passphrase(passphrase, salt)
+            }
+            KdfType::Bcrypt => panic!("BCrypt not implemented")
+        }
+    }
+
+    /// Derive a key given a KDF type, passphrase, and salt
+    async fn load_user_passphrase_key(&self, user: &User, passphrase: &str) -> Result<Aes256Key, ApiCallError> {
+        let Some(salt) = user.salt.as_ref() else {
+            return Err(InternalSdkError { error_message: "Salt is missing from User!".to_string() });
+        };
+        Ok(self.derive_user_passphrase_key(
+            KdfType::Argon2id,
+            passphrase,
+            array_cast_slice(salt, "Vec").map_err(|e|
+            ApiCallError::internal_with_err(e, "Invalid salt")
+            )?).await)
+    }
+
+    /// Initialize a session with given user id and return a new UserFacade
+    async fn init_session(&self, user: &User, user_passphrase_key: Aes256Key) -> Result<UserFacade, LoginError> {
+        let user_facade = UserFacade::new(user.clone());
+
+        user_facade.unlock_user_group_key(user_passphrase_key)
+            .map_err(|e| LoginError::KeyError { error_message: format!("Failed to unlock user group key: {}", e.to_string()) })?;
+        Ok(user_facade)
+    }
+
+    /// Where actual resume session happens.
+    /// Without caching, async login and migration currently.
+    async fn finish_resume_session(&self, credentials: &Credentials) -> Result<UserFacade, LoginError> {
```

Comment:
I have few suggestions for the structure:

 - inline `finish_resume_session()`
 - extract something like `parse_access_key()` either from session or at least from bytes
 - extract `decrypt_passphrase()`
 - you could (but don't have to) return `LoginError` directly from `parse_session_id`  since it's private here

Reply:
I remember us doing it together but now I don't see it?

Reply:
Somehow it's not commited. I make the changes again.